README.md
Fracture (Fork)
=============

![img](/img/img1.png?raw=true "image")

Generative algorithm that makes fracture-like patterns from pre-distributed
randomized seeds.

## Prerequisites

In order for this code to run you must first download and install
repositories:

*    `iutils`: http://github.com/inconvergent/iutils
*    `fn`: http://github.com/inconvergent/fn-python3 (only used to generate
     file names, can be removed in `main.py`.)

## Other Dependencies

The code also depends on:

*    `numpy`
*    `scipy`
*    `python-cairo` (do not install with pip, this generally does not
     work)
*    `pygobject3`
*    `GTK+3`

-----------
http://inconvergent.net

## Fork description

This fork has been developed by Javier Salazar Loyola
([@feandir](https://twitter.com/feandir)), in order to add a front-end
to the algorithm, via a GUI with GTK3 for Python.

## Added features
Export function, in order to save current drawing to PNG or SVG.

## TODO

*    There are some more parameters which can be configured.
*    Figure out if there is a way to change the size of the Cairo
     canvas on the fly, in order to change the size of the picture.
*    Remove dependency on fn-python to name files or modify it in
     order to remove dependency on git
*    Add some threads to handle algorithm execution. Algorithm is slow
     and that affects the interface performance.
*    Move algorithm handling to a new class in order to give a greater
     independence to both algorithm and main window.


ChangeLog
2016-10-14  Javier Esteban Salazar Loyola  <jsalazarloyola@gmail.com>

	* gui/mainWindow.py (MainApp.__createButtons): Set pause button as
	disabled at start and enable it after algorithm started.
	(MainApp.startAlgorithm): Add stop mechanism.
	(MainApp.stopAlgorithm): Change name to a more meaningfull one.

2016-10-12  Javier Esteban Salazar Loyola  <jsalazarloyola@gmail.com>

	* gui/mainWindow.py (MainApp): Add save function to export to PNG
	and SVG.
	(MainApp.__init__): Change default size in order to not have
	useless space.

	* gui/widgets.py: Add ExportDialog class for creating a custom
	save file dialog.

2016-10-11  Javier Esteban Salazar Loyola  <jsalazarloyola@gmail.com>

	* gui/mainWindow.py: Move import to the head of the file.
	(MainApp): Moved buttons to __createButtons method.
	Add Pause toggle button with basic functionality.
	(MainApp.__onExportClicked): Add basic save dialog.
	(MainApp.__init__): Add member to define if algorithm is running.
	(MainApp.__onPauseToggled): Add pause for algorithm execution.

2016-10-07  Javier Esteban Salazar Loyola  <jsalazarloyola@gmail.com>

	* gui/mainWindow.py (MainApp.__createInputs): Add exception for
	handling problems with the image.

2016-10-03  Javier Esteban Salazar Loyola  <jsalazarloyola@gmail.com>

	* gui/mainWindow.py (MainApp.step): Add exception handler for
	errors in the algorithm.

2016-09-30  Javier Esteban Salazar Loyola  <jsalazarloyola@gmail.com>

	* README.md: Add information about the fork and the developed GUI.

	* gui/mainWindow.py (MainApp.setFractures): Add execution of the
	algorithm.
	(MainApp.step): Move the execution to its own function to run when idle.
	(MainApp.__init__): Add connect that allows results to be displayed.
	(MainApp.stoppedAlgorithm): Now fields are reenabled after
	execution and spinner is stopped.

2016-09-29  Javier Esteban Salazar Loyola  <jsalazarloyola@gmail.com>

	* gui/widgets.py (NumberEntry): Add cast to float and int.

	* gui/mainWindow.py (MainApp): Add expose function for drawing the
	canvas.
	(MainApp.show): Add show function to draw the lines through
	drawLines method
	(MainApp.__createInputs): Add missing entries, remove useless
	ones, set default values.

	* gui/widgets.py (NumberEntry): Document the class

2016-09-27  Javier Esteban Salazar Loyola  <jsalazarloyola@gmail.com>

	* gui/widgets.py: Created module for new widgets
	(NumberEntry): Derived class from Gtk.Entry for numerical entries.

	* gui/mainWindow.py (MainApp.__createInputs): Moved inputs from a
	vertical box to a grid.
	(MainApp.__createInputs): Created a dictionary to hold all inputs

2016-09-26  Javier Esteban Salazar Loyola  <jsalazarloyola@gmail.com>

	* gui/mainWindow.py: Created main window draft.
	(MainApp.__createInputs): Added some inputs to modify algorithm.
	TODO: Move the layout to grid

	* mainGui.py: Created main script to handle the GUI.



main.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-




BACK = [1,1,1,1]
FRONT = [0,0,0,0.8]
LIGHT = [0,0,0,0.2]
CYAN = [0,0.5,0.5,0.2]
BLUE = [0,0,1,0.3]


NMAX = 10**6
SIZE = 1200
ONE = 1./SIZE
LINEWIDTH = ONE*1.1

INIT_NUM = 20000
INIT_RAD = 0.45

SOURCE_DST = 2.0*ONE

FRAC_DOT = 0.85
FRAC_DST = 100.*ONE
FRAC_STP = ONE*2
FRAC_SPD = 1.0

FRAC_DIMINISH = 0.997
FRAC_SPAWN_DIMINISH = 0.9


SPAWN_ANGLE = 2.0
SPAWN_FACTOR = 0.2



def show(render,fractures):

  sources = fractures.sources
  alive_fractures = fractures.alive_fractures
  dead_fractures = fractures.dead_fractures

  def draw_sources():
    for i,s in enumerate(sources):
      if i not in fractures.visited:
        render.circle(*s, r=4*ONE, fill=True)

  def draw_lines(fracs):
    for frac in fracs:
      start = frac.inds[0]
      render.ctx.move_to(*sources[start,:])
      for c in frac.inds[1:]:
        render.ctx.line_to(*sources[c,:])
      render.ctx.stroke()

  render.clear_canvas()

  # render.ctx.set_source_rgba(*LIGHT)
  #draw_sources()

  render.ctx.set_source_rgba(*LIGHT)
  render.set_line_width(3*LINEWIDTH)
  draw_lines(alive_fractures+dead_fractures)

  render.ctx.set_source_rgba(*FRONT)
  render.set_line_width(LINEWIDTH)
  draw_lines(alive_fractures+dead_fractures)

  # for f in alive_fractures:
    # for s in sources[f.inds,:]:
      # render.circle(*s, r=2*ONE, fill=False)

def random_uniform_circle(rad, num):

  from numpy.random import random
  from numpy.linalg import norm
  from numpy import array


  while True:
    xy = 0.5-random(size=2)
    if norm(xy)>1.0:
      continue
    r = array([0.5]*2)+xy*rad
    return r



def main():

  from iutils.render import Animate
  from modules.fracture import Fractures

  # from dddUtils.ioOBJ import export_2d as export
  from fn import Fn
  fn = Fn(prefix='./res/',postfix='.2obj')

  F = Fractures(
    INIT_NUM,
    INIT_RAD,
    SOURCE_DST,
    FRAC_DOT,
    FRAC_DST,
    FRAC_STP,
    FRAC_SPD,
    FRAC_DIMINISH,
    FRAC_SPAWN_DIMINISH,
    domain = 'rect'
  )

  print(F.sources.shape)

  # uniform square distribution
  from numpy.random import random
  for _ in range(5):
    F.blow(2, random(size=2))

  # uniform circular distribution
  # for _ in xrange(5):
    # F.blow(3, random_uniform_circle(INIT_RAD, num=1))

  def wrap(render):

    if not F.i % 20:
      show(render,F)
      # vertices, paths = F.get_vertices_and_paths()
      # export('fractures', fn.name(), vertices, lines=paths)
      render.write_to_png(fn.name()+'.png')

    F.print_stats()
    res = F.step(dbg=False)
    n = F.spawn_front(factor=SPAWN_FACTOR, angle=SPAWN_ANGLE)
    print('spawned: {:d}'.format(n))

    # fn = './asdf_{:04d}.png'.format(F.i)
    # render.write_to_png(fn)

    # if not res:
      # vertices, paths = F.get_vertices_and_paths()
      # export('fractures', fn.name(), vertices, lines=paths)

    return res

  render = Animate(SIZE, BACK, FRONT, wrap)
  render.start()


if __name__ == '__main__':

  main()



gui/widgets.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import gi
gi.require_version('Gtk', '3.0')

from gi.repository import Gtk
import re

class NumberEntry(Gtk.Entry):
    """Class for numeric entries in the GUI.

    This class extends the Gtk.Entry class in order to
    take into account for the entries which are required to be
    numbers and not plain text.

    It takes integer and float numbers, but not exponent notation,
    nor long numbers.
    """
    def __init__(self):
        """Default constructor.

        Calls for the Gtk.Entry constructor, and sets functions to run
        when the input changes"""
        # Call base constructor
        Gtk.Entry.__init__(self)
        self.connect('changed', self.on_changed)

        # Regular expression which detects if the text
        # is a valid number. 
        self.rx = re.compile("\d*\.?\d*",re.VERBOSE)

    # Checks if the text is really numbers, otherwise, it will put a blank.
    def on_changed(self,*args):
        """Verifies that the input text is valid."""
        # Finds if the text is really numbers...
        text = self.rx.findall(self.get_text().strip())
        if len(text):
            self.set_text(text[0])
        else:
            self.set_text('')

    def __int__(self):
        """Cast the content to integer

        If the content is float, it returns its floor
        """
        try:
            return int(self.get_text())
        except:
            return int(float(self.get_text()))

    def __float__(self):
        """Cast the content to float"""
        return float(self.get_text())


class ExportDialog(Gtk.FileChooserDialog):
    """Class which handles name finding for exportation of pictures

    This class extends Gtk.FileChooserDialog in order to take into account
    that the fracture images could be exported to PNG or SVG formats.

    It creates the Gtk.FileChooserDialog and sets all its features so the
    main window doesn't have to set it for itself, while also allowing
    further configuration through normal methods of the base class.
    """
    def __init__(self, message, parent=None):
        buttons = (Gtk.STOCK_CANCEL,
                   Gtk.ResponseType.CANCEL,
                   Gtk.STOCK_OPEN,
                   Gtk.ResponseType.OK)
        Gtk.FileChooserDialog.__init__(self, message, parent,
                                       Gtk.FileChooserAction.SAVE,
                                       buttons)

        # Sets filters for images
        self.setFilters()

        # Sets to confirm if an overwrite is to occur
        self.set_do_overwrite_confirmation(True)

        return

    def setFilters(self):
        """Method which sets the default filters for this save dialog"""
        
        """Dictionary that holds the filter list with its associated extensions"""
        self.filters = {}

        # Filter for PNG images
        filterPng = Gtk.FileFilter()
        filterPng.set_name("PNG images")
        filterPng.add_mime_type("image/png")
        self.add_filter(filterPng)
        self.filters[filterPng.get_name()] = ".png"

        # Filter for SVG images
        filterSvg = Gtk.FileFilter()
        filterSvg.set_name("SVG images")
        filterSvg.add_mime_type("image/svg+xml")
        self.add_filter(filterSvg)
        self.filters[filterSvg.get_name()] = ".svg"
        
        return

    def getFileType(self):
        """Gets the file extension from the dictionary"""
        return self.filters[self.get_filter().get_name()]


gui/mainWindow.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import gi, sys
gi.require_version('Gtk', '3.0')

from iutils.render import Render
from gi.repository import Gtk, GdkPixbuf, GObject

import numpy as np
import cairo

from gui.widgets import NumberEntry, ExportDialog

class MainApp(Gtk.Window, Render):
    """Main window for the program

    This is the main view of the program, where every control is present,
    as buttons and text boxes.

    """
    def __init__(self,
                 front, back, light,
                 step=None,
                 width=1000, height=700):
        # Window initialization
        Gtk.Window.__init__(self)

        """If it's currently running"""
        self.stateOn = False
        
        """Front line colors"""
        self.front = front
        """Light color

        Check if this nomenclature is correct
        """
        self.light = light
        
        # Render initialization
        """Size of the drawing area (pixels)"""
        self.dareaSize = height
        """Width of the drawing lines"""
        self.linewidth = 1.1/self.dareaSize
        
        Render.__init__(self, self.dareaSize, front, back)

        border = 5
        #self.inputSpace = width - n - 2*border
        
        # Title and window stuff
        self.set_border_width(border)
        self.set_title("Fractures")
        self.resize(width,height)
        self.set_position(Gtk.WindowPosition.CENTER)

        # Maybe this should be more complex in the future
        self.connect("destroy", self.__quit)

        # Call the input creations
        """Grid that holds all the input data"""
        self.inputGrid = self.__createInputs()
        """Drawing area for the Cairo canvas"""
        self.darea = Gtk.DrawingArea()
        self.darea.set_size_request(self.dareaSize,self.dareaSize)
        self.darea.connect("draw", self.expose)
        """Grid that holds all the buttons"""
        self.buttonsGrid = self.__createButtons()

        # Grid for holding the drawing area (right) and the inputs
        """Grid that holds the input data and the drawing area"""
        self.grid = Gtk.Grid()
        self.grid.add(self.inputGrid)
        self.grid.add(self.darea)
        self.grid.add(self.buttonsGrid)
        
        self.add(self.grid)

        # Show everything on the window
        self.show_all()

    # Function to finish the program
    def __quit(self, button=None):
        Gtk.main_quit()
        
    # Main loop
    def start(self):
        Gtk.main()

    # Method for setting all the inputs in the window
    def __createInputs(self):
        # Grid which holds the elements of the interface
        vgrid = Gtk.Grid()
        vgrid.set_orientation(Gtk.Orientation.VERTICAL)

        # Dictionary which holds the widgets, as the entries
        self.inputsDict = {}
        
        # A nice example picture
        # TODO: get a smaller picture
        try:
            imageName = './img/img1.png'
            imageBuff = GdkPixbuf.Pixbuf.new_from_file_at_scale(imageName,
                                                                width=150,
                                                                height=150,
                                                                preserve_aspect_ratio=True)
        except:
            print("Warning: could not handle the picture. \
            It will not be displayed")
            print("The error:", sys.exc_info()[1])
            
        else:
            image = Gtk.Image()
            image.set_from_pixbuf(imageBuff)
            vgrid.add(image)

        ########################################
        # Inputs as entries
        # Number of sources
        self.sourceNumber = NumberEntry()
        self.sourceNumber.set_text("20000")
        self.inputsDict["sources"] = self.sourceNumber
        vgrid.add(Gtk.Label("Number of sources"))
        vgrid.add(self.sourceNumber)

        # Size of the used region
        self.sizeEntry = NumberEntry()
        self.sizeEntry.set_text("0.45")
        self.inputsDict["size"] = self.sizeEntry
        vgrid.add(Gtk.Label("Size"))
        vgrid.add(self.sizeEntry)
         
        # Minimum distance between sources
        self.distanceEntry = NumberEntry()
        self.distanceEntry.set_text(str(2/self.dareaSize))
        self.inputsDict["distance"] = self.distanceEntry
        vgrid.add(Gtk.Label("Distance"))
        vgrid.add(self.distanceEntry)
        
        ########################################
        # Frequency information
        # Speed of fracture
        self.speedEntry = NumberEntry()
        self.speedEntry.set_text("1.0")
        self.inputsDict["speed"] = self.speedEntry

        self.speedDiminish = NumberEntry()
        self.speedDiminish.set_text("0.997")
        self.inputsDict["speedDiminish"] = self.speedDiminish

        self.spawnDim = NumberEntry()
        self.spawnDim.set_text("0.9")
        self.inputsDict["spawnDim"] = self.spawnDim
        
        vgrid.add(Gtk.Label("Fracture speed"))
        vgrid.add(self.speedEntry)
        vgrid.add(Gtk.Label("Speed diminishing"))
        vgrid.add(self.speedDiminish)
        vgrid.add(Gtk.Label("Fracture spawn diminishing"))
        vgrid.add(self.spawnDim)

        ########################################
        # Spawn angle and factor
        self.spawnFactor = NumberEntry()
        self.spawnFactor.set_text("2.0")
        self.inputsDict["spawnFactor"] = self.spawnFactor

        self.spawnAngle = NumberEntry()
        self.spawnAngle.set_text("0.2")
        self.inputsDict["spawnAngle"] = self.spawnAngle

        vgrid.add(Gtk.Label("Spawn factor"))
        vgrid.add(self.spawnFactor)
        vgrid.add(Gtk.Label("Spawn angle"))
        vgrid.add(self.spawnAngle)
                
        ########################################
        # Domain selection
        domainLabel = Gtk.Label("Domain")
        domainStore = Gtk.ListStore(str, str)
        domainStore.append(['circ', 'Circle'])
        domainStore.append(['rect', 'Rectangle'])

        # Domain combo box
        self.domain = Gtk.ComboBox.new_with_model(domainStore)
        self.domain.connect("changed", self.onSelectionChange)
        renderer = Gtk.CellRendererText()
        self.domain.pack_start(renderer, True)
        self.domain.add_attribute(renderer, "text", 1)
        # Set default value
        self.domain.set_active(0)
        vgrid.add(domainLabel)
        vgrid.add(self.domain)

        # End of function
        return vgrid
        
    def __createButtons(self):
        # Grid which holds the elements of the interface
        vgrid = Gtk.Grid()
        vgrid.set_orientation(Gtk.Orientation.VERTICAL)

        self.buttons = {}
        ########################################
        # Start algorithm button
        startButton = Gtk.Button.new_with_label("Start")
        startButton.connect("clicked", self.startAlgorithm)
        self.runSpinner = Gtk.Spinner()
        self.buttons["start"] = startButton

        vgrid.add(startButton)
        vgrid.add(self.runSpinner)

        # Pause/continue button
        pauseButton = Gtk.ToggleButton("Pause")
        pauseButton.connect("toggled", self.__onPauseToggled, "pause")
        pauseButton.set_sensitive(False)
        vgrid.add(pauseButton)
        self.buttons["pause"] = pauseButton
                    
        # Save button
        saveButton = Gtk.Button("Export picture")
        saveButton.connect("clicked", self.__onExportClicked)
        vgrid.add(saveButton)
        self.buttons["save"] = saveButton
        
        # Close button, just in case
        closeButton = Gtk.Button.new_with_label("Close")
        closeButton.connect("clicked", self.__quit)
        vgrid.add(closeButton)
        self.buttons["close"] = closeButton

        return vgrid

    # When the pause button has been toggled
    def __onPauseToggled(self, button, name):
        if button.get_active():
            # On pause, algorithm is removed from low priority queue
            self.stateOn = False
            button.set_label("Continue")
            print("DEBUG: Algorithm execution paused")
        else:
            # If continue, add algorithm again to low priority queue
            self.stateOn = True
            GObject.idle_add(self.step, self.theFractures, self.fn)
            button.set_label("Pause")
            print("DEBUG: Algorithm execution resumed")
        return

    # Function which handles the exporting of images
    def __onExportClicked(self, button):
        dialog = ExportDialog("Please choose a file", self)
        
        response = dialog.run()
        if response == Gtk.ResponseType.OK:
            print("Save clicked")
            print("File selected: " + dialog.get_filename())
            print("File type: " + dialog.getFileType())
            self.save(dialog.get_filename(), dialog.getFileType())
        elif response == Gtk.ResponseType.CANCEL:
            print("Cancel clicked")
        dialog.destroy()
        return
                
    # What to do when the selection changes in the list of domain options
    def onSelectionChange(self, combo):
        treeiter = combo.get_active_iter()
        if treeiter != None:
            model = combo.get_model()
            self.domainSelection = model[treeiter][0]
            print("Selected:", model[treeiter][0])

    # This function will call the algorithm
    def startAlgorithm(self, button):
        if self.stateOn:
            self.stopAlgorithm()
        else:
            # Remember in the future to stop spinning
            self.runSpinner.start()
            # All the fields will be disabled
            for key in self.inputsDict:
                self.inputsDict[key].set_editable(False)

            self.buttons["pause"].set_sensitive(True)
            self.setFractures()
            self.buttons["start"].set_label("Stop")

        return

    # Function executed when the algorithm reaches its end
    # or the user has invoked an explicit stop
    def stopAlgorithm(self):
        self.stateOn = False
        self.runSpinner.stop()
        self.buttons["start"].set_label("Start")

        # All the fields will be reenabled
        for key in self.inputsDict:
            self.inputsDict[key].set_editable(True)

        self.buttons["pause"].set_sensitive(False)
        return
        
    # Sets the fractures, getting the information from the entries
    # And filling the blanks
    def setFractures(self):
        from modules.fracture import Fractures
        from fn import Fn
        self.fn = Fn(prefix='./res/',postfix='.2obj')

        # These things have been just defined here. I don't know if they should
        # go also as parameters for the front-end.
        fracDot = 0.85
        fracDst = 100./self.dareaSize
        fracStp = 2/self.dareaSize
        
        self.theFractures = Fractures(
            int(self.inputsDict["sources"]),
            float(self.inputsDict["size"]),
            float(self.inputsDict["distance"]),
            fracDot,
            fracDst,
            fracStp,
            float(self.inputsDict["speed"]),
            float(self.inputsDict["speedDiminish"]),
            float(self.inputsDict["spawnDim"]),
            self.domainSelection
        )

        # maybe this to logging.debug?
        print(self.theFractures.sources.shape)
        for _ in range(5):
            self.theFractures.blow(2, np.random.random(size=2))

        # In order to have it running as a low priority process.
        # TODO: -Move this to a thread, instead of idle_add,
        #        and see if it behaves better
        self.stateOn = True
        GObject.idle_add(self.step, self.theFractures, self.fn)

        return

    # Function which advances the algorithm
    def step(self, theFractures, filename):
        if not self.stateOn:
            return False
        
        spawnFactor = float(self.inputsDict["spawnFactor"])
        spawnAngle  = float(self.inputsDict["spawnAngle"])

        fracturesRemain = True
        # Protecting dangerous execution
        try:
            # Show progress every 20 steps
            # (either way it will either hang or last forever
            if not theFractures.i % 20:
                self.show(theFractures)
                self.write_to_png(filename.name()+'.png')

            theFractures.print_stats()
            fracturesRemain = theFractures.step(dbg=False)
            spawned = theFractures.spawn_front(factor = spawnFactor,
                                               angle  = spawnAngle)
            print('spawned: {:d}'.format(spawned))

        except:
            # On exception, prints error message, resets the fields and
            # returns False, in order to release the idle functions of GTK.
            print("Unexpected error:", sys.exc_info()[0])
            print("Traceback:\n", sys.exc_info()[2])
            self.stopAlgorithm()
            return False
        
        # shows the screen
        self.expose()

        # If it reached the end of the algorithm, reenables everything
        if not fracturesRemain:
            self.stopAlgorithm()
        return fracturesRemain

    ############################################################
    # Drawing functions for the canvas
    #
    # Function that draws the picture
    def expose(self, *args):
        # Creates frame for the drawing area
        cairoFrame = self.darea.get_property('window').cairo_create()
        # Copy the drawing from the surface where it's being done
        cairoFrame.set_source_surface(self.sur, 0, 0)
        # Paint the frame
        cairoFrame.paint()
        
        return

    # Function to draw lines between sources. I pressume (addapted from main.py).
    # I don't know how to draw in Cairo, so I'm kind of blindfolded on this part.
    def drawLines(self, fractureSet, sources):
        for frac in fractureSet:
            start = frac.inds[0]
            self.ctx.move_to(*sources[start,:])
            for c in frac.inds[1:]:
                self.ctx.line_to(*sources[c,:])

            self.ctx.stroke()

        return

    def show(self, fractures):
        self.clear_canvas()
        # Draw twice, in order to give some blur effect
        # light, thick lines
        self.ctx.set_source_rgba(*self.light)
        self.ctx.set_line_width(3*self.linewidth)
        self.drawLines(fractures.alive_fractures+fractures.dead_fractures,
                       fractures.sources)

        # strong, thin lines
        self.ctx.set_source_rgba(*self.front)
        self.ctx.set_line_width(self.linewidth)
        self.drawLines(fractures.alive_fractures+fractures.dead_fractures,
                       fractures.sources)

        return

    def save(self, filename, filetype):
        """Function which handles the export of the pictures"""

        # If the file doesn't contain the extension, it's added
        if filename[-4:] != filetype:
            filename += filetype

        # Save PNG
        if filetype==".png":
            # Calls Render.write_to_png
            self.write_to_png(filename)
        # Save SVG
        elif filetype==".svg":
            # Calls mainWindow.write_to_svg
            self.write_to_svg(filename)

        return

    def write_to_svg(self,filename):
        # Create the SVG surface
        # This creates the file on disc automatically)
        svgSurface = cairo.SVGSurface(filename, self.dareaSize, self.dareaSize)
        # Create cairo frame to this surface
        cairoFrame = cairo.Context(svgSurface)
        # Copy surface where the drawing is made to this frame
        cairoFrame.set_source_surface(self.sur, 0, 0)
        # Paint the frame. Now it's saved on disk
        cairoFrame.paint()
        
        return
        
